### PR TITLE
feat(data): insert challenge landscape data into db

### PR DESCRIPTION
**Pull request by vpchung**
The app will now initialize with information curated by the Challenge Landscape:

Database | Table | # of Records
--|--| :--:
challenge_service | `challenge` | 160
challenge_service | `challenge_incentive` | 71
challenge_service | `challenge_platform` | 12
challenge_service | `challenge_submission_type` | 165
organization_service | `challenge_organization` | 233

No records were added or removed from the following tables:

* `challenge_input_data_type`
* `challenge_star`

### TODOs

- [x] add `slug` and `website_url` properties -- I had added these at first to the table creation step, but then realized the properties probably need to be added to the OpenAPI spec as well.  So I have re-commented them out.
- [x] add `slug` and `website_url` values -- this can be done as soon as the table is updated.

### Current blocker
After adding the 160 challenges, I tested the changes and the app worked great!:

Screen Shot 2023-02-03 at 9 34 46 AM

I then tried adding `slug` and `website_url` to the table schema, but thereafter could no longer get the app to work.  I then tried just testing the `challenge-service` by itself and found the following error:

```
2023-02-07 09:57:50.769  WARN [openchallenges-challenge-service,,] 11299 --- [  restartedMain] s.b.e.c.i.ElasticsearchClientFactoryImpl : HSEARCH400073: Hibernate Search will connect to Elasticsearch with authentication over plain HTTP (not HTTPS). The password will be sent in clear text over the network.
2023-02-07 09:57:50.934 ERROR [openchallenges-challenge-service,,] 11299 --- [  restartedMain] o.h.s.e.r.spi.RootFailureCollector       : HSEARCH000521: Hibernate Search encountered a failure during bootstrap; continuing for now to list all problems, but the process will ultimately be aborted.
Context: default backend
Failure:
org.hibernate.search.util.common.SearchException: HSEARCH400080: Unable to detect the Elasticsearch version running on the cluster: HSEARCH400007: Elasticsearch request failed: Connection refused
Request: GET  with parameters {}
Response: (no response)
   ... 
   (truncated for readability)
   ...
Caused by: org.hibernate.search.util.common.SearchException: HSEARCH400007: Elasticsearch request failed: Connection refused
```

Reverted the table schema to its original (without `slug` and `website_url`) and tried testing again but to no avail.  Googling the error message, the connection refusal could be due to a configuration error?

### CODE REVIEW

**Comment by tschaffter** [comment]
> Reverted the table schema to its original (without slug and website_url) and tried testing again but to no avail.

You should be able to add properties and data to the SQL schema of table without affecting the service. The service will ignore the new properties until the Java code is updated. The update is easy to make when adding new properties of 

> org.hibernate.search.util.common.SearchException: HSEARCH400080: Unable to detect the Elasticsearch version running on the cluster: HSEARCH400007: Elasticsearch request failed: Connection refused

This error is to be expected when starting a service that depends on elasticsearch and elasticsearch is still in the process of starting. Is a sign at the end of the log that says that the challenge service successfully connected to elasticsearch after a couple of minutes (I believe that the service keep retrying to connect to elasticsearch).

**Comment by vpchung** [comment]
> You should be able to add properties and data to the SQL schema of table without affecting the service.
Oh, great!

> Is a sign at the end of the log that says that the challenge service successfully connected to elasticsearch
How long would you say this normally takes?  I've had the service up for hours and still cannot connect to elasticsearch.

**Comment by tschaffter** [comment]
It should take about 1-2 minutes to elasticsearch to full start, then it should take less than a minute for the challenge service to index its data into elasticsearch. Does the error occur after a fresh start of the stack (docker artifacts pruned)?

**Comment by vpchung** [comment]
@tschaffter yes, it does 🙃 

I just tried reducing the number of challenges added and it works again!  So either there are too many challenge pre-loaded (which would be odd, seeing that it worked before) or one of the records is wonky after adding in the slugs.

I'll keep sleuthing.

**Comment by vpchung** [comment]
Hmm... I iteratively added challenges until all 160 were re-added and now it's working again... 🤔

**Comment by vpchung** [comment]
More values have been added, as well as a new table that will map challenge → organization + organization's role.  The latest overview of the databases:

Database | Table | # of Records
--|--| :--:
challenge_service | `challenge` | 160
challenge_service | `challenge_incentive` | 71
challenge_service | `challenge_platform` | 12
challenge_service | `challenge_submission_type` | 165
challenge_service | `challenge_organization_role` | 594
organization_service | `challenge_organization` | 236

Additionally, `notebook` has been added to the ENUM for `submission_type` (requested by #1297).

**Comment by vpchung** [comment]
## Blocker

When testing the microservices by themselves, e.g. `nx serve openchallenges-challenge-service`, things are working as expected:

**challenge-service**

Screen Shot 2023-02-07 at 5 14 03 PM


**organization-service**

Screen Shot 2023-02-07 at 5 13 06 PM

When running the app however, challenges still cannot be loaded:

Screen Shot 2023-02-07 at 5 12 28 PM

I double-checked the ports, and they are open and accessible.  The tables are also populated with the expected values.  For some reason, this only seems to to affect challenges.  When I tried looking up one of the newly-added organizations, it worked as expected:

Screen Shot 2023-02-07 at 5 18 31 PM

---

@tschaffter @rrchai any ideas? 😞

**Comment by tschaffter** [comment]
Following our discussion on Slack, I can see the challenges in the search page, at least after merging the `main` branch into this PR branch.

**Comment by vpchung** [comment]
> ...after merging the main branch into this PR branch.

Same result on my end:

Screen Shot 2023-02-08 at 4 16 57 PM

woo!

File: apps/openchallenges/challenge-service/src/main/resources/db/migration/V1.0.0__create_tables.sql [hunk 1]
```diff
@@ -29,10 +29,11 @@ CREATE TABLE `challenge_input_data_type`
 CREATE TABLE `challenge`
 (
     `id`                    bigint(20) NOT NULL AUTO_INCREMENT,
+    `slug`                  varchar(255) NOT NULL UNIQUE,
     `name`                  varchar(255) DEFAULT NULL,
     `headline`              varchar(80),
     `description`           varchar(280) NOT NULL,
-    `avatar_url`            varchar(255),
+    -- `avatar_url`            varchar(255),
```
Comment (tschaffter) [inline apps/openchallenges/challenge-service/src/main/resources/db/migration/V1.0.0__create_tables.sql:36]: Why did you remove the avatar?

File: apps/openchallenges/challenge-service/src/main/resources/db/migration/V1.0.0__create_tables.sql [hunk 2]
```diff
@@ -29,10 +29,11 @@ CREATE TABLE `challenge_input_data_type`
 CREATE TABLE `challenge`
 (
     `id`                    bigint(20) NOT NULL AUTO_INCREMENT,
+    `slug`                  varchar(255) NOT NULL UNIQUE,
```
Comment (tschaffter) [inline apps/openchallenges/challenge-service/src/main/resources/db/migration/V1.0.0__create_tables.sql:32]: Note: It's possible that the slug may end up not being unique. This would be OK if we retrieve challenges by their ID, e.g. with the app route `/challenge/{id}/{slug}`.

File: apps/openchallenges/challenge-service/src/main/resources/db/migration/V1.0.0__create_tables.sql [hunk 3]
```diff
@@ -53,6 +51,20 @@ CREATE TABLE `challenge`
 -- KEY                `FKk9w2ogq595jbe8r2due7vv3xr` (`account_id`),
 -- CONSTRAINT `FKk9w2ogq595jbe8r2due7vv3xr` FOREIGN KEY (`account_id`) REFERENCES `banking_core_account` (`id`)
 
+-- challenge_organization_role definition
+
+CREATE TABLE `challenge_organization_role`
```
Comment (tschaffter) [inline apps/openchallenges/challenge-service/src/main/resources/db/migration/V1.0.0__create_tables.sql:56]: What do you think about renaming this table `challenge_contribution_role` to make it more flexible. This would open the door to non-organization contribution, e.g. org that participated to a challenge (future feature?).

File: apps/openchallenges/db-cli/README.md [hunk 4]
```diff
@@ -7,23 +7,23 @@ instance.
 
 Check the code style.
 
-    nx lint challenge-db-cli
+    nx lint openchallenges-db-cli
```
Comment (tschaffter) [inline apps/openchallenges/db-cli/README.md:10]: Thanks for taking the time to update the doc!

File: apps/openchallenges/challenge-service/src/main/resources/db/migration/V1.0.0__create_tables.sql (hunk 2 above)
Comment (vpchung) [reply apps/openchallenges/challenge-service/src/main/resources/db/migration/V1.0.0__create_tables.sql:32]: Are the new routes already implemented?  I will keep this as unique until then.

Comment (tschaffter) [reply apps/openchallenges/challenge-service/src/main/resources/db/migration/V1.0.0__create_tables.sql:32]: We should have the new routes implemented later this month.

Comment (tschaffter) [reply apps/openchallenges/challenge-service/src/main/resources/db/migration/V1.0.0__create_tables.sql:32]: cc @rrchai see my initial comment about above about the route design for challenges.

Comment (vpchung) [reply apps/openchallenges/challenge-service/src/main/resources/db/migration/V1.0.0__create_tables.sql:32]: Constraint has been removed.

**Comment by tschaffter** [comment]
@vpchung Is this PR still blocked? Do you want me to review it again?

**Comment by vpchung** [comment]
Oh yes, sorry, it's been ready 🙃 

Just requested another review.

File: apps/openchallenges/challenge-service/src/main/resources/db/migration/V1.0.0__create_tables.sql [hunk 5]
```diff
@@ -71,7 +82,7 @@ CREATE TABLE `challenge_incentive`
 CREATE TABLE `challenge_submission_type`
 (
     `id`                    int NOT NULL AUTO_INCREMENT,
-    `name`                  ENUM('container_image', 'prediction_file', 'other'),
+    `name`                  ENUM('container_image', 'prediction_file', 'notebook', 'other'),
```
Comment (tschaffter) [inline apps/openchallenges/challenge-service/src/main/resources/db/migration/V1.0.0__create_tables.sql:85]: This line should already be in `main`. I'm wondering why GH is still flagging it.

**Review by tschaffter — APPROVED** [review]
